src/ethereum/vm/instructions.py

```
@@ -13,16 +13,40 @@
 """
 
 
-from ..base_types import U256
+from functools import partial
+from typing import cast
+
+from ..base_types import U255_CEIL_VALUE, U256, U256_MAX_VALUE, Uint
+from ..utils import signum
 from . import Evm
 from .gas import (
+    GAS_EXP,
+    GAS_EXP_BYTE,
+    GAS_LOW,
+    GAS_MID,
     GAS_STORAGE_CLEAR_REFUND,
     GAS_STORAGE_SET,
     GAS_STORAGE_UPDATE,
     GAS_VERY_LOW,
     subtract_gas,
 )
-from .stack import pop, push
+from .stack import pop, pop_multiple, push
+
+#
+# Arithmetic Operations
+#
+
+
+def stop(evm: Evm) -> None:
+    """
+    Stop further execution of EVM code.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+    """
+    evm.pc = Uint(len(evm.code))
```

This is a somewhat surprising way (to me) to implement stop. Is this how it's specified?

---

Maybe we raise here and catch it gracefully?

---

The yellowpaper doesn't specify the implementation for this, and this seemed the easiest way to do this. Further, I didn't want to raise an error for this, as it didn't seem like an error to me. It's just another opcode. Let me know your thoughts on this.

---

> it didn't seem like an error to me

I would agree. I think we should use exceptions for the "exceptional" halts in the yellowpaper.

---

I think the best approach for clarity would be to add a `running` flag to the `Evm` class, and change the [interpreter loop](https://github.com/quilt/eth1.0-specs/blob/efe41d6687b70ed84d1dc4b5d80c6c0c5ad98750/src/ethereum/vm/interpreter.py#L86) to something more like this:

```python
while evm.running:
    op = evm.code[evm.pc]
    op_implementation[op](evm)
    evm.pc += 1

    if evm.pc >= len(evm.code):
        evm.running = False
```

`stop` would just set `evm.running` to `False`, which seems like the obvious implementation to me.